crates/holochain_zome_types/src/block.rs

```diff
@@ -0,0 +1,201 @@
+    DOS,
+}
+
+/// Reason why we might want to block an IP.
+#[derive(Clone, serde::Serialize, Debug)]
+pub enum IPBlockReason {
+    /// Classic DOS.
+    DOS,
+}
+
+// @todo this is probably wrong.
+type NodeId = [u8; 32];
+
+/// The type to use for identifying blocking ipv4 addresses.
+type IpV4 = std::net::Ipv4Addr;
+
+/// Target of a block.
+/// Each target type has an ID and associated reason.
+#[derive(Clone, Debug)]
+pub enum BlockTarget {
+    /// Some cell did bad at the happ level.
+    Cell(CellId, CellBlockReason),
+    /// Some node is playing silly buggers.
+    Node(NodeId, NodeBlockReason),
+    /// An entire college campus has it out for us.
+    IP(IpV4, IPBlockReason),
+}
+
+#[derive(Debug, serde::Serialize, Clone)]
+pub enum BlockTargetId {
+    Cell(CellId),
+    Node(NodeId),
+    IP(IpV4),
+}
+
+impl From<BlockTarget> for BlockTargetId {
+    fn from(block_target: BlockTarget) -> Self {
+        match block_target {
+            BlockTarget::Cell(id, _) => Self::Cell(id),
+            BlockTarget::Node(id, _) => Self::Node(id),
+            BlockTarget::IP(id, _) => Self::IP(id),
+        }
+    }
+}
+
+#[cfg(feature = "rusqlite")]
+impl ToSql for BlockTargetId {
+    fn to_sql(&self) -> rusqlite::Result<ToSqlOutput<'_>> {
+        Ok(rusqlite::types::ToSqlOutput::Owned(
+            holochain_serialized_bytes::encode(&self)
+                .map_err(|e| rusqlite::Error::ToSqlConversionFailure(Box::new(e)))?
+                .into(),
+        ))
+    }
+}
+
+#[derive(Debug, serde::Serialize, Clone)]
+pub enum BlockTargetReason {
+    Cell(CellBlockReason),
+    Node(NodeBlockReason),
+    IP(IPBlockReason),
+}
+
+#[cfg(feature = "rusqlite")]
+impl ToSql for BlockTargetReason {
+    fn to_sql(&self) -> rusqlite::Result<ToSqlOutput<'_>> {
+        Ok(rusqlite::types::ToSqlOutput::Owned(
+            holochain_serialized_bytes::encode(&self)
+                .map_err(|e| rusqlite::Error::ToSqlConversionFailure(Box::new(e)))?
+                .into(),
+        ))
+    }
+}
+
+impl From<BlockTarget> for BlockTargetReason {
+    fn from(block_target: BlockTarget) -> Self {
+        match block_target {
+            BlockTarget::Cell(_, reason) => BlockTargetReason::Cell(reason),
+            BlockTarget::Node(_, reason) => BlockTargetReason::Node(reason),
+            BlockTarget::IP(_, reason) => BlockTargetReason::IP(reason),
+        }
+    }
+}
+
+/// Represents a block.
+/// Also can represent an unblock.
+/// NOT serializable and NOT pub fields by design. `try_new` MUST be the only
+/// entrypoint to build a `Block` as this enforces that the start/end times are
+/// valid according to invariants the SQL queries rely on to avoid corrupting the
+/// database.
+#[derive(Clone, Debug)]
+pub struct Block {
+    /// Target of the block.
+    target: BlockTarget,
+    /// Start time of the block. None = forever in the past.
+    start: Timestamp,
+    /// End time of the block. None = forever in the future.
+    end: Timestamp,
```

Comment:
Why not Option? I think that would make sense. Otherwise, update the docs to not say None.

```suggestion
    /// Start time of the block. None = forever in the past.
    start: Option<Timestamp>,
    /// End time of the block. None = forever in the future.
    end: Option<Timestamp>,
```

Reply:
I also wonder if the time window thing is YAGNI at this stage? The main use case right now is failing validation, and that's always a permablock..

Reply:
@maackle one of the main use cases is also blocking based on transient things like paying a subscription or holding an NFT

Reply:
@maackle it isn't option because in this case None just becomes an alias of Timestamp::MIN or Timestamp::MAX, so i've made that explicit rather than implicit